include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h

```diff
@@ -1470,8 +1470,11 @@ class ThreadPoolTempl : public onnxruntime::concurrency::ExtendedThreadPoolInter
     void SetBlocked(std::function<bool()> should_block,
                     std::function<void()> post_block) {
       std::unique_lock<std::mutex> lk(mutex);
-      assert(GetStatus() == ThreadStatus::Spinning);
-      status.store(ThreadStatus::Blocking, std::memory_order_relaxed);
+      auto old_status = status.exchange(ThreadStatus::Blocking, std::memory_order_seq_cst);
+      if (old_status != ThreadStatus::Spinning) {
+        // Encountered a logical error
+        throw std::runtime_error("The state transition is not allowed");
+      }
```

Comment:
What about assert + [read barrier](https://learn.microsoft.com/en-us/cpp/intrinsics/readbarrier?view=msvc-170) + atomic store?

Reply:
With the exchange function, we can do all three things in a single function call.  If we split the read and write, the assert statement isn't meaningful .

Reply:
It seems bit more clear to do it that way.

And yes, as I understand it, the assert is no longer necessary for correctness hence it can be just an assert and not this handrolled version.